Multi Producer Single Consumer channel

huntc opened the pull request:
An MPSC inspired by Tokio and Crossbeam. The MPSC targets the mutexes we presently have i.e. those for single-core processors at this time. That said, the design is agnostic and should be able to target multi-core processors. 

The allocation of the channel’s buffer is inspired by the const generic parameters that Heapless provides.

Sample usage as per a new "mpsc" example:

```rust
enum LedState {
    On,
    Off,
}

static CHANNEL: Forever<Channel<WithNoThreads, LedState, 1>> = Forever::new();

#[embassy::task(pool_size = 1)]
async fn my_task(sender: Sender<'static, WithNoThreads, LedState, 1>) {
    loop {
        let _ = sender.send(LedState::On).await;
        Timer::after(Duration::from_secs(1)).await;
        let _ = sender.send(LedState::Off).await;
        Timer::after(Duration::from_secs(1)).await;
    }
}

#[embassy::main]
async fn main(spawner: Spawner, p: Peripherals) {
    let mut led = Output::new(p.P0_13, Level::Low, OutputDrive::Standard);

    let channel = CHANNEL.put(Channel::new());
    let (sender, mut receiver) = mpsc::split(channel);

    spawner.spawn(my_task(sender)).unwrap();

    // We could just loop on `receiver.recv()` for simplicity. The code below
    // is optimized to drain the queue as fast as possible in the spirit of
    // handling events as fast as possible. This optimization is benign when in
    // thread mode, but can be useful when interrupts are sending messages
    // with the channel having been created via with_critical_sections.
    loop {
        let maybe_message = match receiver.try_recv() {
            m @ Ok(..) => m.ok(),
            Err(TryRecvError::Empty) => receiver.recv().await,
            Err(TryRecvError::Closed) => break,
        };
        match maybe_message {
            Some(LedState::On) => led.set_high().unwrap(),
            Some(LedState::Off) => led.set_low().unwrap(),
            _ => (),
        }
    }
}
```

The above code can be run via `cargo run --bin mpsc` from the examples/nrf folder (and does run!):

```
     Running `probe-run --chip nRF52840_xxAA embassy/target/thumbv7em-none-eabi/release/mpsc`
(HOST) INFO  flashing program (8.88 KiB)
(HOST) INFO  success!
```

Fixes #29 

TODO as a result of feedback:

* [x] providing a const type param indicating the max concurrent sends (and closes) seems ok - see https://github.com/embassy-rs/embassy/pull/226#issuecomment-873521502
* [x] if the max concurrent sends is reached then we log via the debug! macro - see commit 42a00ec4d4f309a94fa9c97b263dac3ccebe7fe2
* [x] think on passing in a mutex trait to the channel shared between CriticalSectionMutex and ThreadModeMutex

lulf commented:
FWIW, there is an existing channel impl we have for drogue-device actor framework https://github.com/drogue-iot/drogue-device/blob/main/device/src/kernel/channel.rs , which I suspect have similar requirements to yours. I would be keen on finding a common api/impl that could be in embassy. I think it would be a useful utility for communicating between embassy tasks.

Dirbaio commented on embassy-extras/src/mpsc.rs:
```diff
@@ -0,0 +1,257 @@
+/// # Panics
+///
+/// Panics if the buffer capacity is 0.
+pub fn channel<'a, T>(buf: &'a mut [u8]) -> (Sender<T>, Receiver<T>) {
```
I don't think this is possible to implement without `alloc`. `Sender` and `Receiver` would need a lifetime param so they can borrow the buffer (`Sender<'a, T>`). 

We probably need to store more state than just the buf (wakers, fill amount..) too, so we probably need a Channel type to hold it (and have Sender/Receiver borrow the Channel instead of the buf)

huntc replied:
Absolutely. I’ve not fleshed out the structs at all yet.

huntc replied:
Btw the approach to construction and retention of this buffer slice will be as be the existing circular buffer we have.

huntc replied:
> I don't think this is possible to implement without `alloc`.

I think I've now caught up with you. :-)

The reasons for `alloc`, so far as I can see are so that Sender objects can be cloned between tasks. Is this what you were thinking @Dirbaio?

huntc replied:
No longer using the circular buffer.

huntc commented:
> FWIW, there is an existing channel impl we have for drogue-device actor framework https://github.com/drogue-iot/drogue-device/blob/main/device/src/kernel/channel.rs , which I suspect have similar requirements to yours. I would be keen on finding a common api/impl that could be in embassy. I think it would be a useful utility for embassy tasks to share state.
> 
> FWIW I think the interface matches the one I would expect of an async channel too.

It’s similar in terms of interface. I think the obvious difference is, as per Tokio, the recv and send functions here are async. The try_send and try_recv aren’t - which is more like yours. Wdyt?

lulf commented:
> > FWIW, there is an existing channel impl we have for drogue-device actor framework https://github.com/drogue-iot/drogue-device/blob/main/device/src/kernel/channel.rs , which I suspect have similar requirements to yours. I would be keen on finding a common api/impl that could be in embassy. I think it would be a useful utility for embassy tasks to share state.
> > FWIW I think the interface matches the one I would expect of an async channel too.
> 
> It’s very similar in terms of interface. I think the obvious difference is, as per Tokio, the recv and send functions here are using the async keyword.

That's true, but at the inner layer you'd still need to implement your own future to handle wakeup etc.

Is there a particular reason for accepting a byte slice instead of just wrapping a a "heapless" mpsc queue?

Another thought: could this have been a useful addition to the heapless crate?

huntc commented:
> That's true, but at the inner layer you'd still need to implement your own future to handle wakers registration, notification when channel has items available etc.

Absolutely, and your code would certainly be inspirational there, I'm sure.

> Is there a particular reason for accepting a byte slice instead of just wrapping a a "heapless" mpsc queue?

I was inspired by Embassy's `ring_buffer` module there.

> Another thought: could this have been a useful addition to the heapless crate?

Possibly. I was wondering if they'd be anything in the implementation that could leverage other parts of Embassy in the same manner that Tokio's channels do. However, we could also provide a nostd channel as its own crate entirely, similar to [Crossbeam](https://crates.io/crates/crossbeam).

huntc commented:
I've been exploring the use of critical sections for this channel abstraction. There's a precedent for the use of critical sections elsewhere in the codebase, recognising that they are only suitable for single-core scenarios. I've also relocated the MPSC into the src/util folder, alongside mutex and signal; seems therefore appropriate.

I've written some tests although I notice that the project doesn't support tests. I'd like to discuss that. Clippy also finds a few problems.

There's more to do here; in particular, hooking up the send/receive abstractions with the underlying private channel that I introduce.

Dirbaio commented:
Very interesting!

In many usecases you're sending data between two tasks that are on the same executor, so running at the same priority level. In that case you shouldn't need any mutexing, because these tasks can't preempt each other.

futures_intrusive does something interesting: it makes it generic over the mutex implementation. You can instantiate a channel with a "noop mutex" which makes it not Send (so it's only useful to send data within the same prio level) or you can use a real mutex (for example based on critical sections) which makes it Send so you can use it across priority levels.

IMO it's not ideal to force using CS's for uses that don't need them, so maybe we should do something similar to the above so we can have both?

huntc commented:
I suppose I could use a mutex. But then the mutex uses a CS under the hood I think. I should have mentioned that I went down this path as the nRF event queue does exactly the same.

btw I took a look a futures_intrusive. I really wanted to use it, but I couldn’t fathom it so easily.

huntc commented:
Actually, I’m reading your comments again. Seems like I missed the point you were making. 

I’ll absorb and then reply!

huntc commented:
> IMO it's not ideal to force using CS's for uses that don't need them, so maybe we should do something similar to the above so we can have both?

I've been thinking about this a bit. I think you're right in terms of not exposing the CS, as per the `Signal` example. The code so far actually doesn't expose the CS at the public level. The private `Channel` implementation does use one though, and it is important to do so so that no other interrupts fire for synchronised operations. The approach I'm taking "feels" correct then, in the same way that the `Signal` implementation also uses critical sections.

huntc commented:
> Is there a particular reason for accepting a byte slice instead of just wrapping a a "heapless" mpsc queue?

I've now adopted the same approach to heapless. Much nicer. Thanks for the question.

huntc commented:
> Very interesting!
> 
> In many usecases you're sending data between two tasks that are on the same executor, so running at the same priority level. In that case you shouldn't need any mutexing, because these tasks can't preempt each other.
> 
> futures_intrusive does something interesting: it makes it generic over the mutex implementation. You can instantiate a channel with a "noop mutex" which makes it not Send (so it's only useful to send data within the same prio level) or you can use a real mutex (for example based on critical sections) which makes it Send so you can use it across priority levels.
> 
> IMO it's not ideal to force using CS's for uses that don't need them, so maybe we should do something similar to the above so we can have both?

The API now looks something like this:

```rust
        let (s, r): (Sender<u32, 3>, Receiver<u32, 3>) = channel();
        assert!(s.try_send(1).is_ok());
        assert_eq!(r.try_recv().unwrap(), 1);
```

As of commit 1cfeea2411be1dc909ee990ac7e1316b211a04a3, I've not put together the sharing of the channel correctly. I thought I'd maintain a pointer to a refcell of the channel in the sender, and retain the channel in the receiver. However, things are late here I don't have things totally working. Happy to receive some thoughts on how best to share the channel object though.

huntc commented:
Per commit adecab0a89111a4def9be1f13ac10d524ced46dd, the API now looks like this:

```rust
        let mut channel = Channel::<u32, 3>::new();
        let (sender, receiver) = split(&mut channel);
        assert!(sender.clone().try_send(1).is_ok());
        assert_eq!(receiver.try_recv().unwrap(), 1);
```

In essence, all that is left for me to do is implement some channel registration/deregistration functionality.

huntc commented:
Almost functionally complete as at commit e2bf1135e68c204face82421ab174682d81fcac5. All that remains is to implement the async `recv` and async `send` with the respective `try_` methods. I shall also introduce an mpsc example illustrating the usage from futures.

I'm starting to think that the mpsc could be in its own repo as it doesn't use anything Embassy-specific. It might be nice to retain the Embassy-enabled example though as part of this PR though. Any thoughts on this?

huntc commented:
I've just pushed an example of what the Embassy integration looks like. 

Firstly, I'm not liking this declaration:

```rust
static mut CHANNEL: mpsc::Channel<LedState, 1> = mpsc::Channel::new();
```

Static mutables are unsafe this requiring in the main function:

```rust
    let (sender, mut receiver) = unsafe { mpsc::split(&mut CHANNEL) };
```

Please see: https://github.com/embassy-rs/embassy/pull/226/files#diff-2c2da7e397c20984db93cf4b9aa3c4ddf66899ad18dd2a0262ed1cf1d0a3ab4c

huntc commented:
> I've just pushed an example of what the Embassy integration looks like.
> 
> Firstly, I'm not liking this declaration:
> 
> ```rust
> static mut CHANNEL: mpsc::Channel<LedState, 1> = mpsc::Channel::new();
> ```
> 
> Static mutables are unsafe thus requiring in the main function:
> 
> ```rust
> let (sender, mut receiver) = unsafe { mpsc::split(&mut CHANNEL) };
> ```
> 
> Please see: https://github.com/embassy-rs/embassy/pull/226/files#diff-2c2da7e397c20984db93cf4b9aa3c4ddf66899ad18dd2a0262ed1cf1d0a3ab4c

Following on from this, the static mutable approach is at least as per Heapless (see the `XS` example): https://docs.rs/heapless/0.7.1/heapless/

huntc commented:
> > I've just pushed an example of what the Embassy integration looks like.
> > Firstly, I'm not liking this declaration: ...

The appearance of `unsafe` has now been superseded by the use of Embassy's `Forever` type.

lulf reviewed (commented):
I think its looking pretty good @huntc! I don't have time to review in detail right now, but I wonder about the multiple producer aspect of this.

From what I can tell, there can only be a single sender waiting on a channel, as there is only one waker stored. In order for multiple senders to wait, would it not need to store a list of wakers?

FWIW I think just single producer single consumer is sufficient for many use cases, just wanted to check.

huntc commented:
> From what I can tell, there can only be a single sender waiting on a channel, as there is only one waker stored. In order for multiple senders to wait, would it not need to store a list of wakers?
> 
This is an oversight on my part. I should store a collection of sender wakers. Thanks for the quick inspection and comment!

huntc commented:
> This is an oversight on my part. I should store a collection of sender wakers. Thanks for the quick inspection and comment!

Should I go for an array that limits the number of concurrent senders?

huntc commented:
> Should I go for an array that limits the number of concurrent senders?

One idea could be to remove clone from sender and have a constant number of senders returned by split…

huntc commented:
> One idea could be to remove clone from sender and have a constant number of senders returned by split…

Another idea is to specify the number of concurrent sends in the type signature in the same way that the size of the channel buffer is conveyed e.g.:

```rust
static CHANNEL: Forever<mpsc::Channel<LedState, 3, 1>> = Forever::new();
```

...conveys that we require a buffer size of 3 but that we have a concurrent sender factor of 1. Any sender exceeding this concurrency factor would, perhaps, result in a new sender error of "Busy" or something... Thoughts?

huntc commented:
And my last idea for the time being is to use a linked list for the sender wakers. Perhaps the best idea. -)

huntc commented:
> Another thought: could this have been a useful addition to the heapless crate?

@lulf FYI I’ve now asked this question at https://github.com/japaric/heapless/issues/218

huntc commented:
As per commit 42a00ec4d4f309a94fa9c97b263dac3ccebe7fe2, I've thought more on "providing a const type param indicating the max concurrent sends (and closes) seems ok". I've effectively assumed the max concurrent sends as 1 for simplicity. This will result in the potential for spins, but given that `recv` should be called from within a tight loop, I'm wondering if the spinning will result in a significant problem. To that end, I've placed a trace log so that we encourage people to report back on this. The mitigation is to introduce the const type param and a circular buffer to hold send wakers; although falling back onto this "wake the old sender" logic if the buffer fills.

huntc commented:
> > From what I can tell, there can only be a single sender waiting on a channel, as there is only one waker stored. In order for multiple senders to wait, would it not need to store a list of wakers?
> 
> This is an oversight on my part. I should store a collection of sender wakers. Thanks for the quick inspection and comment!

Now handled. I'm presently avoiding the complexity of a list of wakers. See: https://github.com/embassy-rs/embassy/pull/226#issuecomment-873521502.

huntc commented:
I've been playing around with the, " think on passing in a mutex trait to the channel shared between CriticalSectionMutex and ThreadModeMutex" task, and here's an illustration of what it looks like in the example:

```rust
enum LedState {
    On,
    Off,
}

static CHANNEL: Forever<Channel<WithThreadModeOnly<LedState, 1>, LedState, 1>> =
    Forever::new();

#[embassy::task(pool_size = 1)]
async fn my_task(
    sender: Sender<'static, WithThreadModeOnly<LedState, 1>, LedState, 1>,
) {
    loop {
        let _ = sender.send(LedState::On).await;
        Timer::after(Duration::from_secs(1)).await;
        let _ = sender.send(LedState::Off).await;
        Timer::after(Duration::from_secs(1)).await;
    }
}

#[embassy::main]
async fn main(spawner: Spawner, p: Peripherals) {
    let mut led = Output::new(p.P0_13, Level::Low, OutputDrive::Standard);

    let channel = CHANNEL.put(Channel::with_thread_mode_only());
    let (sender, mut receiver) = split(channel);

    spawner.spawn(my_task(sender)).unwrap();

    loop {
        match receiver.recv().await {
            Some(LedState::On) => led.set_high().unwrap(),
            Some(LedState::Off) => led.set_low().unwrap(),
            _ => (),
        }
    }
}
```

I haven't yet found a way to reduce that type signature down. Something like this might read nicer:

```rust
Channel<WithThreadModeOnly<LedState, 1>>> 
```

...but again, I'm not there yet.

huntc commented:
> I've been playing around with the, " think on passing in a mutex trait to the channel shared between CriticalSectionMutex and ThreadModeMutex" task, and here's an illustration of what it looks like in the example:
> 

This is now resolved. Quite happy with how the types look:

```rust
static CHANNEL: Forever<Channel<WithThreadModeOnly, LedState, 1>> = Forever::new();
```

huntc commented:
All TODOs completed. Ready for a thorough review @Dirbaio!

huntc commented:
> > Another thought: could this have been a useful addition to the heapless crate?
> 
> Possibly. I was wondering if they'd be anything in the implementation that could leverage other parts of Embassy in the same manner that Tokio's channels do. However, we could also provide a nostd channel as its own crate entirely, similar to [Crossbeam](https://crates.io/crates/crossbeam).

Just as a follow-up on this one. While there's still scope for this channel to live outside of Embassy, it has ended up introducing a `Mutex` trait that is implemented for the existing CS and thread mode mutexes. It is beginning to feel correctly placed here. Further, I feel that I might be using the channel along with the core Embassy crate with [my actor crate](https://github.com/titanclass/stage). In fact, once this PR is in then I'll write a dispatcher for my actors that will run on Embassy and have its request/reply `ask` functionality enabled.

lulf reviewed (commented):
@huntc Great stuff. I've been trying to replace my own channel implementation with this one in drogue-device, and just a few observations on things that would improve usage:

* Having a single ChannelError would be more ergonomic when wrapping errors (otherwise one need a variant for each error type)
* Remove async-keyword from send, and instead return SendFuture (making it a public type). I feel this is more ergonomic to use when wanting to pass the returned future off to something else. 
* Same as above, but with receive, requires introducing a RecvFuture similar to SendFuture.

huntc commented:
> @huntc Great stuff. I've been trying to replace my own channel implementation with this one in drogue-device, and just a few observations on things that would improve usage:
> 
> …

Thanks for testing this out!

I will have a good think about your suggestions. Meanwhile, do you know that the API I have is that of Tokio’s?

lulf commented:
> > @huntc Great stuff. I've been trying to replace my own channel implementation with this one in drogue-device, and just a few observations on things that would improve usage:
> > …
> 
> Thanks for testing this out!
> 
> I will have a good think about your suggestions. Meanwhile, do you know that the API I have is that of Tokio’s?

I haven't really used Tokio APIs in much detail, and I've not used async APIs enough to really say what the best approach is. Just that I personally find it easier when wrapping types, to work out type signatures if I see a named type than an implicitly generated type from the async keyword.

huntc commented:
As of commit b81e8008d0e27fb4a0e24f8d0297f04341b6d362, I've introduced a `ChannelLike` trait so that I can reduce the types on sender and receiver e.g.: 

```rust
pub struct Sender<'ch, T> {
    channel: &'ch UnsafeCell<dyn ChannelLike<T>>,
}
```

This improves the user's code a little, and the channel implementation a lot.

I'd prefer not to have the user pass in the `UnsafeCell` as they now do this:

```rust
    let channel = CHANNEL.put(UnsafeCell::new(Channel::with_thread_mode_only()));
```

An `Arc` would be ideal. :-)

thalesfragoso commented on embassy/src/util/mpsc.rs:
```diff
@@ -0,0 +1,881 @@
+    fn close(&mut self) {
+        let state = &mut self.state;
+        self.mutex.lock(|_| {
+            state.receiver_waker.take().map(|w| w.wake());
```
I think it's a bit weird to use map just for the side-effects, maybe a `if let` is better here ?

huntc replied:
Changed. Thanks.

huntc commented:
I've introduced my own `UnsafeCell` and named it `ChannelCell` with the aim to share channels between the sender and receiver, and also remove the unsafe connotation. ChannelCell is safe within the context of being used via its senders and receivers.

I was also trying to constraint the type that ChannelCell would accept but I'm missing some Foo. Inspiration is welcome there.

Commit 0595f1850468ebecbb636de46b5721ca9184efc1

huntc commented:
> I was also trying to constrain the type that ChannelCell would accept but I'm missing some Foo. Inspiration is welcome there.

ChannelCell is now constrained to ChannelLike.

lulf commented on embassy/src/util/mpsc.rs:
Wdyt about making this type public, along with a pub fn new() that accepts a Sender instance and a message to be sent? This should allow for the use case where the future type needs to be known in advance.

huntc replied:
Why have a public `new` though?

huntc replied:
The type should be opaque to reduce the surface area, right?

lulf replied:
@huntc To avoid needing to use the Sender.send() function for cases that need to know the future type. For instance, so one can use the channel in code like this:

```rust
struct MyStruct {
   future: SendFuture<'static, MyMessage>
}

fn foo(sender: Sender<'static, MyMessage>) -> MyStruct {
     MyStruct {
        future: SendFuture::new(sender.clone(), MyMessage)
    }
}
```

I'm not sure how I can store the future returned from Sender.send() for later use otherwise without needing alloc.

Dirbaio commented on embassy/src/util/mpsc.rs:
```diff
@@ -0,0 +1,905 @@
+    }
+}
+
+impl<'ch, M, T, const N: usize> Future for Receiver<'ch, M, T, N>
```
It's somewhat strange that `Receiver` impls `Future`. Futures are only supposed to run once. Maybe making a `ReceiveFuture` would make sense.

huntc replied:
> It's somewhat strange that `Receiver` impls `Future`. 

There are other examples of this in Embassy, e.g. `Timer`.

> Futures are only supposed to run once. Maybe making a `ReceiveFuture` would make sense.

Whoops - mistake. Commit ec18ae1f8836815773a5ba744c65404bc64f0af4 fixes this.

Dirbaio commented on embassy/src/util/mpsc.rs:
```diff
@@ -0,0 +1,905 @@
+    M: Mutex<Data = ()>,
+{
+    sender: Sender<'ch, M, T, N>,
+    message: UnsafeCell<T>,
```
Why is this UnsafeCell needed?

huntc replied:
In the event of an error, the message is also returned (same as Tokio's):

```rust
            Err(TrySendError::Closed(m)) => Poll::Ready(Err(SendError(m))),
```

huntc replied:
Would could strengthen perhaps with RefCell, but I "know" there's just one copy of the message to be passed around i.e. it is a bug otherwise.

Dirbaio replied:
Okay, it's not simply unneeded, it's incorrect:

- When you create the SendFuture, the `T` is moved into it.
- If SendFuture completes successfully, the `T` should be moved from it into the channel. The `T` should NOT be dropped.
- If SendFuture is canceled (by dropping it), the `T` should be dropped.

The current code is *always* dropping the T on SendFuture drop. On successful send, it's doing a bitwise-copy of the T, store that T in the channel buffer, but *still keep a copy in SendFuture* which gets dropped when SendFuture is dropped.

It's UB to drop the same object twice, it causes big problems if it's eg `Box<T>`. 

I think you need an `Option<T>`...



Related: SendFuture will never return SendError::Full, it blocks if it is full. Maybe it's better to have separate `SendError` and `TrySendError` enums?

huntc replied:
> Okay, it's not simply unneeded, it's incorrect:

Removed as per commit ae572e94307d226b3ea45e878724694d2528affa

huntc replied:
> Related: SendFuture will never return SendError::Full, it blocks if it is full. Maybe it's better to have separate SendError and TrySendError enums?

I already have a separate `SendError` which is only ever returned by this future when the channel has been closed...

Dirbaio commented on embassy/src/util/mpsc.rs:
```diff
@@ -0,0 +1,905 @@
+    closed: bool,
+    receiver_registered: bool,
+    senders_registered: u32,
+    receiver_waker: Option<Waker>,
```
These should use WakerRegistration. With the embassy executor, it makes it so wakers are just 1 word instead of 2, and also avoids unnecessary cloning which is a bit more efficient in executors that refcount wakers.

huntc replied:
Done as per commit 766a3958018b8fd66f34dfe2381678afe683feb7

Dirbaio commented on embassy/src/util/mpsc.rs:
```diff
@@ -0,0 +1,905 @@
+        let receiver_waker = None;
+        let senders_waker = None;
+        ChannelState {
+            buf,
```
it's nicer to write all the initial values inline: `ChannelState { buf: [xx], read_pos: 0, write_pos: 0...}`

huntc replied:
Done as per commit 7acf14d9ffbc11216e7a504ec9b7f52639e8f660

Dirbaio commented on embassy/src/util/mpsc.rs:
````diff
@@ -0,0 +1,905 @@
+    /// // once we have a channel, obtain its sender and receiver
+    /// let (sender, receiver) = mpsc::split(&channel);
+    /// ```
+    pub const fn with_critical_sections() -> Self {
````
Would it be possible to have just one `new()` constructor? The channel already knows the type of mutex it should use from the generic params.

huntc replied:
I wanted to do that too. Here's what I get though:

```
error[E0034]: multiple applicable items in scope
   --> embassy/src/util/mpsc.rs:779:78
    |
779 |         static mut CHANNEL: Channel<WithCriticalSections, u32, 3> = Channel::new();
    |                                                                              ^^^ multiple `new` found
    |
note: candidate #1 is defined in an impl for the type `mpsc::Channel<mutex::CriticalSectionMutex<()>, T, N>`
   --> embassy/src/util/mpsc.rs:453:5
    |
453 |     pub const fn new() -> Self {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #2 is defined in an impl for the type `mpsc::Channel<mutex::ThreadModeMutex<()>, T, N>`
   --> embassy/src/util/mpsc.rs:477:5
    |
477 |     pub const fn new() -> Self {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #3 is defined in an impl for the type `mpsc::Channel<mutex::NoopMutex<()>, T, N>`
   --> embassy/src/util/mpsc.rs:498:5
    |
498 |     pub const fn new() -> Self {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Dirbaio replied:
`lock_api`'s  `RawMutex` solves this with the [`const INIT`](https://docs.rs/lock_api/0.4.4/lock_api/trait.RawMutex.html#associatedconstant.INIT). WIth it, `new()` can generically construct a mutex instance without knowing which type it is.

huntc replied:
I tried a few things here. I'm likely missing some foo. Would you be able to provide a bit more of an example?

huntc replied:
Commit bd6cd3471d888e7d72a132f13b53aabf09414586 attends to this. Thanks for the code @Dirbaio!

Dirbaio commented on embassy/src/util/mpsc.rs:
```diff
@@ -0,0 +1,905 @@
+where
+    M: Mutex<Data = ()>,
+{
+    fn get(&self) -> &mut Self {
```
converting `&self` to `&mut self` like this is unsound: you may not have two `&mut` to the same data, or a `&mut` and a `&` to the same data.

huntc replied:
Yes. This is effectively equivalent to UnsafeCell though, right? Only the senders and receiver may access this method and they cannot outlive the channel. Each of the senders and the one receiver must point back to the channel.

Dirbaio replied:
Yes and no.  First, if this method requires "special care" to be used correctly to not cause UB, it should be `unsafe fn`.

Second, you may say "okay, it's `unsafe fn`, and Sender/Receiver take care to use it correctly (only call it while holding the mutex and only once, so there's never two `&mut` concurrently), it's *still* UB.

Why?

The compiler is allowed to assume a function taking `&` never mutates the data if the data is not behind UnsafeCell. For example:

```rust
struct Foo { bar: u32 }

let foo = Foo { bar: 42 }
do_stuff(foo);
whatever(foo.bar);
```

The compiler is allowed to optimize the last line to `whatever(42)`, because it's allowed to assume `do_stuff` can't mutate `foo.bar` because it takes a `&Foo`. It is UB for `do_stuff` to use `unsafe` to convert `&Foo` to `&mut Foo`. It can cause Bad Stuff to happen, like broken optimizations as seen here.

`UnsafeCell` is "special", it warns the compiler that "interior mutability" can happen, which will prevent these optimizations *for data inside the `UnsafeCell` ONLY*. ALL mutated data must be inside the `UnsafeCell`. 

Also, it's STILL not allowed to cast `&` to `&mut`. Even if everything in the struct is inside an `UnsafeCell`. You're only allowed to get a `&mut` to stuff *inside the `UnsafeCell`*, not to the outer struct. Castinv `&` to `&mut` is ALWAYS UB, no exceptions.

This is the reason why all types that allow for interior mutability always end up using `UnsafeCell` internally: Cell, RefCell, Mutex...

Dirbaio replied:
Further reading: 

- https://doc.rust-lang.org/reference/interior-mutability.html
- https://doc.rust-lang.org/reference/behavior-considered-undefined.html

Most notably, from the last link:

> - Breaking the pointer aliasing rules. &mut T and &T follow LLVM’s scoped noalias model, except if the &T contains an UnsafeCell<U>.
> - Mutating immutable data. All data inside a const item is immutable. Moreover, all data reached through a shared reference or data owned by an immutable binding is immutable, unless that data is contained within an UnsafeCell<U>.

huntc replied:
I never realised that `UnsafeCell` was special to the compiler. Thanks for the detailed explanations. I'm going to overhaul this area and look to use `RefCell` instead.

Dirbaio replied:
RefCell is not zero-cost though. I think it's possible to get away with just Cell/UnsafeCell.

huntc replied:
> RefCell is not zero-cost though. I think it's possible to get away with just Cell/UnsafeCell.

Ok

huntc replied:
I've attended to this via commit 29bfa6b7b3d1a3108382d8b4cec10617c24878cc. I used `UnsafeCell`. `Cell` would require the use of `take` and `swap` to perform the interior mutation, which I don't think makes sense for us here as we want multiple threads to be able to access the internal channel, but locked when required.

To this end, I've introduced a `lock` method on the channel in place of the UB `get` that I had before. This ensures that all access to the channel is synchronised. A new `ChannelCell` has been introduced to support our requirements around interior mutability. The channel's methods have been moved to the `ChannelState` type, which also greatly simplifies the code.

Dirbaio commented on embassy/src/util/mpsc.rs:
```diff
@@ -0,0 +1,905 @@
+    read_pos: usize,
+    write_pos: usize,
+    full: bool,
+    closing: bool,
```
Why is `closing` needed? I think on close you can immediately set `closed = true`, but on recv check if there's available data anyway.

huntc replied:
I think what I have is more explicit. The channel can be closed explicitly prior to performing a `recv`, in which case you need to continue to consume. However, you shouldn't consume if closed...

Dirbaio replied:
It's extra code and 1 extra byte of state for IMO no benefit. We should try to make it as efficient as possible.

huntc replied:
Here's a scenario to consider:

1. there is one sender
2. sender sends
3. the sender is dropped
4. the receiver then receives

When a sender is dropped, it will de-register from the channel. When there are no senders left, the channel is put into a `closing` state. The channel is not closed at this stage. If it was then the receiver would not be able to receive what was sent.

The `receiver_receives_given_try_send_async` test covers this scenario.

Dirbaio replied:
You can still achieve that without the `closing` state. Receive does the following:
- If buffer not empty, pop a value, return it.
- Else, if closed return Closed error
- Else return Empty error

Channel is immediately closed, but receivers won't see it until after draining all the items.

huntc replied:
I'm convinced. :-) Commit bed15ada14959186bcb308bd981489b6d7426773 removed the `closing` flag.

Dirbaio commented on embassy/src/util/mutex.rs:
```diff
@@ -78,3 +105,30 @@ pub fn in_thread_mode() -> bool {
+
+/// A "mutex" that does nothing and cannot be shared between threads.
+pub struct NoopMutex<T> {
+    inner: UnsafeCell<T>,
```
this could be simply `inner: T`

huntc replied:
Nice one! Commit f5199cbbf3df9146e13baa06d5b63863c8794e15 attends to this.

Dirbaio commented on embassy/src/util/mpsc.rs:
```diff
@@ -0,0 +1,905 @@
+    type Output = Option<T>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.try_recv() {
```
This can lead to a race condition. `try_recv` and `set_receiver_waker` both lock the channel state independently, in between them it is unlocked and another thread could affect the state. This is a problem, for example:

- This thread calls try_recv, it returns Empty
- Another thread pushes a value. No waker is registered yet, so it doesn't wake anything.
- This thread calls `set_receiver_waker`, registering the waker even though the channel is no longer empty.
- End result: current task doesn't pop the value out, and waits forever instead.

Easiest way to avoid this is to ensure each "operation" such as this poll only locks the mutex once, so that it's atomic.

huntc replied:
Well spotted! Fixed as per your suggestion of an atomic operation in commit 196f555e61571055fd8e111bbbf3358c895fe01c

Dirbaio reviewed (changes requested):
Looks great overall, however there are a few soundness issues.

Thanks for working on this :)

huntc commented on embassy/src/util/mpsc.rs:
```diff
@@ -0,0 +1,905 @@
+}
+
+// Safe to pass the sender around
+unsafe impl<'ch, M, T, const N: usize> Send for Sender<'ch, M, T, N> where M: Mutex<Data = ()> {}
```
TBH, these were a bit copypasta from other channel impls. The suggestion here is that we can remove them?

huntc replied:
If I remove these then I get:

```
error: future cannot be sent between threads safely
   --> embassy/src/util/mpsc.rs:815:14
    |
815 |             .spawn(async move {
    |              ^^^^^ future created by async block is not `Send`
    |
    = help: within `mpsc::Channel<mutex::CriticalSectionMutex<()>, u32, 3_usize>`, the trait `Sync` is not implemented for `UnsafeCell<u32>`
note: captured value is not `Send`
   --> embassy/src/util/mpsc.rs:816:22
```

...and more in my tests. This is associated with UnsafeCell being used a bit later for the SendFuture. I'll describe why that's needed further down.

Dirbaio replied:
Interesting, maybe it's not possible to have rust auto-impl these then? No idea.

Either way, *always* impling Send/Sync is wrong, it depends on the mutex type. These should have a `where M: Sync` bound.

huntc replied:
Commit 2d24965dc56d5cff943c35b24be5e589e4f11412 attends to this.

huntc commented on embassy/src/util/mpsc.rs:
````diff
@@ -0,0 +1,905 @@
+/// };
+/// ```
+pub fn split<M, T, const N: usize>(
+    channel: &Channel<M, T, N>,
````
> Since this takes `&Channel`, it's actually possible to call it multiple times on the same channel. I'd argue it doesn't really "split" then, just "create a sender and receiver".
> 

It does actually split, although it should only be allowed to call it once. The question here for me is how can we achieve that?

Tokio and others do so by providing a `channel` function that creates the channel on the heap and returns the sender and a single receiver. We don't have the heap of course, and our channel buffer is created separately using a const generic size.

> One alternative is to require `&mut`, though I imagine this will impact usability.
> 

I'm not sure how this can be done.

> Another alteriative is to embrace it's possible to create senders/receivers on demand, and convert `split` into `sender` and `receiver` methods in Channel? In this case the channel would really be MPMC, maybe we should just call it "Channel" then, not specify the kind.

I was really going for MPSC from both: (a) the perspective of being able to optimise for that; and (b) being able to reason that a channel's value is consumed from just one place.

I'm not sure things are optimised for a single consumer now, but that is still my design goal.

huntc replied:
Other than advertising this an MPMC, I can't think of anything we can do to enforce the single consumer at compile time. Right now, I think the best we can do re. MPSC is to provide the runtime check as we do.

huntc replied:
It occurred to me that, by splitting a borrow on the channel, I can have the compiler enforce MPSC. I've now taken care of this as per commit 37710dba94d515465706d2700ac3b9ac5124a38e.

In essence, the sender and receiver structures reference the ChannelCell. Furthermore, for the receiver, it mutably borrows a `receiver_consumed` field, thus preventing another receiver existing for the same channel at the same time.

huntc commented:
> Looks great overall, however there are a few soundness issues.
> 
> Thanks for working on this :)

My pleasure! I'm learning heaps! I've addressed all of your feedback if not only with a few more bits of info for some. Thanks for the detailed review.

Dirbaio commented on embassy/src/util/mpsc.rs:
```diff
@@ -0,0 +1,887 @@
+        {
+            Ok(..) => Poll::Ready(Ok(())),
+            Err(TrySendError::Closed(m)) => Poll::Ready(Err(SendError(m))),
+            Err(TrySendError::Full(..)) => {
```
Even worse, this contains a `T` that will get dropped, so you may be dropping the T a toooon of times.

huntc replied:
This should now be resolved

Dirbaio commented on embassy/src/util/mpsc.rs:
```diff
@@ -0,0 +1,887 @@
+            .sender
+            .channel
+            .get()
+            .try_send_with_context(unsafe { self.message.get().read() }, Some(cx))
```
It would be simpler to have `poll_send` in `Channel` and just call it, so you can avoid converting the error here, and the issues with moving the T around.

huntc replied:
I no longer have issues copying T around given commit ae572e94307d226b3ea45e878724694d2528affa. Do you think this still applies? I've got some nice re-use going with `try_send_with_context`.

Dirbaio replied:
Yup, looks correct now!

Dirbaio replied:
Making everything a single function might help in moving it less, but we can optimize that later. Let's get a working, correct version merged first! Marking as resolved

Dirbaio commented on embassy/src/util/mpsc.rs:
```diff
@@ -0,0 +1,887 @@
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        if self.sender.channel.get().is_closed_with_context(Some(cx)) {
```
Why does `close` have to block until close is "done"? In other channels I've seen, `close` completes instantly, even if the channel still has queued items. Receivers still receive these items, and receive the "closed" error when there are no more queued items.

It seems more intuitive to me: `send` doesn't wait for receivers to receive the item, so `close` shouldn't wait for receivers to receive the close.

huntc replied:
Note that the receiver's `close` method completes immediately, whereas the sender's `closed` method completes when the receiver is dropped. Its Tokio API doc states, "This allows the producers to get notified when interest in the produced values is canceled and immediately stop doing work.".

Dirbaio commented on embassy/src/util/mpsc.rs:
```diff
@@ -0,0 +1,887 @@
+}
+
+struct ChannelState<T, const N: usize> {
+    buf: [MaybeUninit<UnsafeCell<T>>; N],
```
`MaybeUninit` never drops the contents when dropped. This means dropping a Channel that still has queued items doesn't drop the items. This should have a Drop impl that drops the queued items. (*Only* the queued ones, which may be less than N!)

huntc replied:
Nice catch!

huntc replied:
Fixed with commit 29639c5ebc1d575224957a95e66eb590738da949

huntc commented on embassy/src/util/mpsc.rs:
```diff
@@ -0,0 +1,905 @@
+}
+
+// Safe to pass the receiver around
+unsafe impl<'ch, M, T, const N: usize> Send for Receiver<'ch, M, T, N> where M: Mutex<Data = ()> {}
```
Should now be resolved.

Dirbaio reviewed (approved):
Let's do this :rocket: 

Thank you @huntc for your hard work! This is awesome :)

huntc commented:
> Thank you @huntc for your hard work! This is awesome :)

Thank you so much for the time you've invested into its review!
